crypto/dsa/dsa_check.c

```diff
@@ -19,14 +19,25 @@
 #include "dsa_local.h"
 #include "crypto/dsa.h"
 
-int dsa_check_params(const DSA *dsa, int *ret)
+int dsa_check_params(const DSA *dsa, int *codes)
 {
-    /*
-     * (2b) FFC domain params conform to FIPS-186-4 explicit domain param
-     * validity tests.
-     */
-    return ossl_ffc_params_FIPS186_4_validate(dsa->libctx, &dsa->params,
-                                              FFC_PARAM_TYPE_DSA, ret, NULL);
+    FFC_PARAMS params = {0};
+    int ret;
+
+    if (dsa->params.seed != NULL)
```

Comment:
I think you have broken FIPS by fixing the default provider case...
This needs to be #ifndef FIPS_MODULE

Reply:
This also means the buildinfo needs to change so this file is compiled twice.

Reply:
@slontis - I disagree I've broken anything. The full FIPS check does not have any meaning with NULL seed. Or in other words - if you do not pass seed you cannot expect to perform the full FIPS check. On the other hand, even with the FIPS module it is useful to have at least some check for the parameters you've passed in.

Reply:
The scenario I am talking about is when you call 'validate_params' in FIPS mode.
If there is no seed then it is an error - It can't just do a partial check and say that it is good.

This is part of the reason for moving to using named groups for dh.

Reply:
See #14146 for a way forward..

Reply:
@slontis So you're basically saying that openssl pkeyparam -check should do the quick check and not the full check? Or what?

IMO overloading the function that is called by openssl pkeyparam -check with meaning FIPS specific validate_params call in the FIPS module does not make any sense, because that can simply never pass.

Reply:
Please note that to base the decision whether the FIPS validation or the simple check is requested on the dsa->params.seed being non NULL is fine because the seed will never be NULL if the FIPS validation is properly requested.

And the named groups are not allowed for DSA anyway so this is not really relevant here.

Reply:
if pkeyparam -check is always going to fail then
So we either do
(1) Just do the quick check OR
(2) We fail in fips mode.

> dsa->params.seed being non NULL is fine because the seed will never be NULL if the FIPS validation is properly requested

How do you jump to this conclusion? If you load a dsa key and call validate in FIPS mode and it is a fips 186-4 key then it should fail if the extra params used for generation (not just the seed) are not present. Without those extra parameters there is no assurance that the key is correct.